File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/RemoteServiceApi.java
package com.heaton.funnyvote.data;

import com.heaton.funnyvote.retrofit.Server;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by chiu_mac on 2016/12/6.
 */

public class RemoteServiceApi {
    private static final String TAG = RemoteServiceApi.class.getSimpleName();
    public static final String USER_TYPE_FACEBOOK = "fb";
    public static final String USER_TYPE_GOOGLE = "google";
    public static final String USER_TYPE_TWITTER = "twitter";

    //Retrofit
    Retrofit retrofit;

    public Server.UserService getUserService() {
        return userService;
    }

    public Server.VoteService getVoteService() {
        return voteService;
    }

    public Server.PromotionService getPromotionService() {
        return promotionService;
    }

    Server.UserService userService;
    Server.VoteService voteService;
    Server.PromotionService promotionService;
    private static RemoteServiceApi INSTANCE;

    public RemoteServiceApi() {
        retrofit = new Retrofit.Builder().baseUrl(Server.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        userService = retrofit.create(Server.UserService.class);
        voteService = retrofit.create(Server.VoteService.class);
        promotionService = retrofit.create(Server.PromotionService.class);
    }

    public static RemoteServiceApi getInstance() {
        if (INSTANCE == null) {
            synchronized (RemoteServiceApi.class) {
                if (INSTANCE == null) {
                    INSTANCE = new RemoteServiceApi();
                }
            }
        }
        return INSTANCE;
    }

}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/RemoteServiceApi.kt
package com.heaton.funnyvote.data

import com.heaton.funnyvote.retrofit.Server

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by chiu_mac on 2016/12/6.
 */

class RemoteServiceApi {

    //Retrofit
    internal var retrofit: Retrofit = Retrofit.Builder().baseUrl(Server.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

    var userService: Server.UserService
        internal set
    var voteService: Server.VoteService
        internal set
    var promotionService: Server.PromotionService
        internal set

    init {
        this.userService = retrofit.create(Server.UserService::class.java)
        this.voteService = retrofit.create(Server.VoteService::class.java)
        this.promotionService = retrofit.create(Server.PromotionService::class.java)
    }

    companion object {
        @JvmField
        val TAG: String? = RemoteServiceApi::class.java.simpleName
        const val USER_TYPE_FACEBOOK = "fb"
        const val USER_TYPE_GOOGLE = "google"
        const val USER_TYPE_TWITTER = "twitter"
        private var INSTANCE: RemoteServiceApi? = null

        @JvmStatic
        fun getInstance(): RemoteServiceApi {
            return INSTANCE ?: RemoteServiceApi().apply {
                INSTANCE = this
            }
        }
    }

}
